plugins/importexport/onix30/Onix30ExportPlugin.inc.php

```diff
@@ -101,18 +104,67 @@ function display($args, $request) {
 				]);
 				$templateMgr->display($this->getTemplateResource('index.tpl'));
 				break;
+			case 'exportSubmissionsBounce':
+				if (!$request->checkCSRF()) throw new Exception('CSRF mismatch!');
+				$json = new JSONMessage(true);
+				$json->setEvent('addTab', array(
+					'title' => __('plugins.importexport.native.results'),
+					'url' => $request->url(null, null, null, array('plugin', $this->getName(), 'export'), array('selectedSubmissions' => $request->getUserVar('selectedSubmissions'), 'csrfToken' => $request->getSession()->getCSRFToken())),
+				));
+				header('Content-Type: application/json');
+				return $json->getString();
 			case 'export':
+				$onixDeployment = new Onix30ExportDeployment($context, $user);
+
 				$exportXml = $this->exportSubmissions(
 					(array) $request->getUserVar('selectedSubmissions'),
-					$request->getContext(),
-					$request->getUser()
+					$context,
+					$user,
+					$onixDeployment
 				);
+
+				$problems = $onixDeployment->getWarningsAndErrors();
+				$foundErrors = $onixDeployment->isProcessFailed();
+
+				if ($exportXml) {
+					$dateFilenamePart = date($exportFileDatePartFormat);
+					$this->writeExportedFile($exportXml, $onixDeployment->getContext(), $dateFilenamePart);
+
+					$templateMgr->assign('exportPath', $dateFilenamePart);
+				}
+
+				$templateMgr->assign('validationErrors', $onixDeployment->getXMLValidationErrors());
+
+				$templateMgr->assign('errorsAndWarnings', $problems);
+				$templateMgr->assign('errorsFound', $foundErrors);
+				$templateMgr->assign('onixPlugin', $this);
+
+				$json = new JSONMessage(true, $templateMgr->fetch($this->getTemplateResource('resultsExport.tpl')));
+				header('Content-Type: application/json');
+				return $json->getString();
+			case 'downloadExportFile':
+				if (!$request->checkCSRF()) throw new Exception('CSRF mismatch!');
+				$downloadPath = $request->getUserVar('downloadFilePath');
+				
+				$date = DateTime::createFromFormat($exportFileDatePartFormat, $downloadPath);
```

Comment:
I think you could skip the double-validating (both here and in `getExportFilename`) if the new parameter to `getExportFilename` was a `?DateTime`  object. Then you could validate here when calling `DateTime::createFromFormat`, and within `getExportFilename`, just format to a string when the parameter is provided.